Fixes #12630 - Facets registration

ShimShtein opened the pull request:
First part of #2665.
This part only adds an ability to register your facet.

ShimShtein committed:
Fixes #12630 - Facets registration

dLobatog commented on app/models/host_facet_base.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  # Use this method to populate host's fields based on fact values exposed by the importer.
+  # You can populate fields in the associted host's facets too.
```
Typo, associated (associted)

dLobatog commented on app/models/host_facet_base.rb:
```diff
@@ -0,0 +1,43 @@
+
+  # Use this method to populate host's fields based on fact values exposed by the importer.
+  # You can populate fields in the associted host's facets too.
+  def self.populate_fields_from_facts(host, importer, type, proxy_id)
```
`FactImporter` already has `@host` as an instance variable, and `type` simply comes from `facts[:_type]`. `@facts` is also an instance variable in `FactImporter`, so are `host` and `type` necessary here? 

Same goes for proxy_id, at most make it default to nil, I'm thinking that at least for ForemanAnsible it's not exactly necessary, we should be able to allow to import facts without a proxy.

ShimShtein replied:
> so are host and type necessary here?

`@host` is private in the importer, so I don't have access to it in the facet. I give it to the facet, so the facet will decide what properties it wants to change in host object (for example build the facet).
`type` exists in the [original  method signature](https://github.com/theforeman/foreman/blob/develop/app/models/host/managed.rb#L465), so I didn't want to change it.

The whole `proxy_id` issue needs refactoring in foreman regardless my PR.

dLobatog commented on app/models/host_facet_base.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  # Change attributes that will be sent to an facet based on inherited values from the hostgroup.
+  def self.inherited_attributes(hostgroup, facet_attributes)
```
Shouldn't the default be `hostgroup.facet_attributes`? As I understand it, it'll just default to return the child attributes when calling inherited_attributes

ShimShtein replied:
`facet_attributes` is the part of facet's attributes from `apply_inherited_attributes`. The implementation looks like this:

``` Ruby
def apply_inherited_attributes(attributes, initialized = true)
#...
  Facets.registered_facets.values.each do |facet_config|
    facet_attributes = attributes["#{facet_config.name}_attributes"]
    facet_attributes = facet_config.model.inherited_attributes(new_hostgroup, facet_attributes)
    attributes["#{facet_config.name}_attributes"] = facet_attributes if facet_attributes
  end
#...
```

dLobatog commented on app/models/host_facet_base.rb:
```diff
@@ -0,0 +1,43 @@
+  end
+
+  # Specify any smart proxy id's used by this facet.
+  def smart_proxy_ids
```
Why smart proxy _id's_ vs smart proxies (objects)? The latter option seems easier to use.

ShimShtein replied:
In `Host::Managed` there is already a [method](https://github.com/theforeman/foreman/blob/develop/app/models/host/managed.rb#L743) `#smart_proxies`.

dLobatog commented on app/services/facets.rb:
```diff
@@ -0,0 +1,24 @@
+  def register(facet_model, facet_name = nil, &block)
+    entry = Facets::Entry.new(facet_model, facet_name)
+
+    entry.instance_eval(&block) if block_given?
```
You can make this slightly faster by not passing `&block` as an argument, check out http://mudge.name/2011/01/26/passing-blocks-in-ruby-without-block.html . I don't think it matters a lot though, but a comment above the method giving an example of facet that would use `block` would help. I can't remember any right now.

ShimShtein replied:
Anything except very basic initialization is done via the block:

``` Ruby
  Facets.register(PuppetFacet) do
    extend_model PuppetHostExtensions
    add_helper PuppetFacetHelper
    add_tabs :puppet_tabs
    api_view :list => 'api/v2/puppet_facets/base', :single => 'api/v2/puppet_facets/single_host_view'
    template_compatibility_properties :environment_id, :puppet_proxy_id, :puppet_ca_proxy_id
  end
```

As part of the documentation effort, I will add this example too.

dLobatog commented on app/services/facets/entry.rb:
```diff
@@ -0,0 +1,67 @@
+      raise ArgumentError, "facet_model must be a class" unless facet_model.is_a? Class
+
+      facet_name ||= to_name(facet_model)
+      raise ArgumentError, "facet_name must be a symbol" unless facet_name.is_a? Symbol
```
When would this not be a symbol if `to_name` calls `.to_sym` at the end?

ShimShtein replied:
If it is specified explicitly

dLobatog commented on app/services/facets/entry.rb:
```diff
@@ -0,0 +1,67 @@
+    def api_docs(param_group, controller, description = nil)
+      raise ArgumentError, "param_group must be a symbol" unless param_group.is_a? Symbol
+      raise ArgumentError, "controller must be a class" unless controller.is_a? Class
+      raise ArgumentError, "description must be a string" unless (description || '').is_a? String
```
If description defaults to `''` I think you can save that `||` conditional.

ShimShtein replied:
In the implementation I am using the fact that the description is actually empty:

``` Ruby
#...
desc = facet_config.api_param_group_description || N_("Parameters for host's %s facet" % facet_config.name)
#...
```

I can move this code to here actually, but IMHO it's not a trivial place to look for user visible strings.

dLobatog commented on app/services/facets/entry.rb:
```diff
@@ -0,0 +1,67 @@
+module Facets
+  class Entry
+    attr_accessor :name, :model, :helper, :extension,
```
If all of these variables are `attr_accessors`, what do we gain with the type checking below? In the end I could call `Facets::Entry.last.helper = 'not a module'` and it wouldn't complain. That's probably fine actually.

I'm not sure what to think about argument checking in Ruby, but I don't know any option that could be clearly better. I would suggest for this class to not worry about types, and let the Facet implementations worry about whether `:tabs` is an array, or whatever. 

The reason I propose this is so that any Facet could literally use anything as `tabs`, `helper`, etc... and all the `rescues`, exceptions and so forth would fail or raise an exception where they are called, not this early. It'd be more flexible and doesn't force us to define a very strongly typed type of Facet::Entry.

Summing up my point is that the responsibility for making sure a Facet works with the arguments provided during Facet registration lies on the person writing the Facet, not the "Facet framework".

@domcleal @jlsherrill any ideas?

jlsherrill replied:
I would tend to agree as even these checks aren't completely fool proof.  I could pass a model class that isn't actually active record, same for the controller check.  You could check the base class but i feel that is heading down the a rabbit whole. 

I would prefer to see the facet API well documented here with good usage examples.  I think that is more valuable than validation checks.

ShimShtein replied:
I wanted to create public readers, private writers pattern here. I'll fix it to `attr_reader` instead.
I agree that good documentation needed here, but not everybody reads it. I prefer adding at least basic integrity check here, since this is the place where the error actually is. I want to minimize the cases when the user get an error at some random point in code (let's say in API controller) that doesn't tell the user too much.

dLobatog commented on app/services/facets/entry.rb:
```diff
@@ -0,0 +1,67 @@
+      :tabs,
+      :compatibility_properties
+
+    @compatibility_properties = []
```
This should be within `initialize`?

dLobatog commented on app/services/facets/entry.rb:
```diff
@@ -0,0 +1,67 @@
+      self.api_param_group_description = description
+    end
+
+    def template_compatibility_properties(*property_symbols)
```
Please leave a comment above the methods (or attr_accessor definition, depending what we end up doing) explaining what exactly they are for. Some are clear like api_docs after reading them, but some others are a bit more contrived (like this one)

dLobatog commented:
@ShimShtein Thanks for breaking this up. 

I didn't check tests as it might be better to not force this kind of contract for Facets and making Facet::Entry a very lightweight class with just attr_accessors and a few helper methods. I'm not completely sure which path to follow here so if you could get some attention from other devs to this problem (Facet::Entry type checking) that'd be good, and we can decide. @ohadlevy ?

In case we resort to simple voting, :-1: to making Facet::Entry check types, and :+1: to letting Facet writers figure out that on their own Facets, where these arguments passed during registration can fail (or not).

Other than that thing about Facet::Entry, I think only a few comments & simple stuff should be fixed, great job :hammer:

lzap commented:
I agree with some docs, can you start a wiki page or something with examples? This patch alone does not make lot of sense and I wonder if this breakage is not little bit too far, but I am fine with this. Good start.

ShimShtein commented:
- Documentation added
- attr_accessor changed
- typo fixed

ohadlevy commented:
[test]

dLobatog commented:
@ShimShtein I've been looking at the Puppet facet implementation for some time. Just to get things going, let's see if we agree in a few things:

1) **Type checking `Facet::Entry` arguments** - I don't see the point, it reduces flexibility for facet writers a lot. We don't type check plugin's registration or other additions either, we let them fail when they have to fail. I think this would allow people to pass whatever they want to api_docs/view etc and handle that as they see fit.

By the way, thanks for the docs on `Facet::Entry`, that's surely going to be helpful!

2) **Can we do with a leaner `HostFacetBase`**?  Doesn't the current one make a lot of assumptions? 

e.g: `info` - do all facets need to modify the ENC? do all hosts even need an ENC representation? If i'm not using Puppet I don't see why would I use it. 
`smart_proxies_ids` - same thing, all facets need to use smart_proxies?
etc... 

The only one I could see being almost universally used would be `inherited_attributes`, and that seems restricting too. Shouldn't facets be inherited from hostgroup to host (not asking for you to implement that here)?   

``` ruby
class HostFacetBase < ActiveRecord::Base
  self.abstract_class = true
  belongs_to_host
end
```

Last, kind of a nitpick, can it live in `app/models/host_facets/base.rb`? We follow a similar model with nics and host statuses.

3) After this, what **would be the next step**? I think a PR to update `Host::Managed` to use the kind of stuff that I see on `HostFacetBase` now would be the wrong course to go vs simply extracting any facet (if there's something simpler than Puppet, that's fine). I think generalizations such as moving stuff to `HostFacetBase` should happen as implementations come in & not before. Let me know what do you think could go next.

4) It doesn't look like facets registration needs a lot of testing so don't worry if the tests associated to this PR look meagre.  

Sorry for the 1 month delay on coming back to this. We should be close to merging though :smile: , depending on how we/others see points 1 & 2 above.

ShimShtein commented:
- Rebased
- Removed typechecks (We can always add those checks later, if we see that the users are confused)
- Renamed `HostFacetBase` to `HostFacets::Base`

> Can we do with a leaner HostFacetBase

Since the system doesn't know a-priori which facets extend which parts of the host, in each extension point in `Host::Managed` it asks all facets to contribute data ([example](https://github.com/ShimShtein/foreman/blob/puppet-facet/app/models/host/managed.rb#L469)). I can either make a call to each facet, and facets that do not override some extension point will return a no-op result (this is what happens now) or I can ask each facet does it implements this extension point. I tried the latter approach with `responds_to?`, but @domcleal didn't liked it. Personally I don't like the second approach too, because in current way facet developers have a full list of extension points that are available to them.

> I think a PR to update Host::Managed to use the kind of stuff that I see on HostFacetBase now would be the wrong course to go vs simply extracting any facet

My only simpler candidate is Katello's subscriptions/content facet, but they are plugins. It won't solve the no usage in the same commit problem.
Any facet extraction will litter the commit with tons of facet-specific code, so we will have the same bulky commit as in puppet. Another reason for not doing it, is that we will not be able to separate UI/API extensions from the model extensions, which will clutter the PR even more.
If you can think of some low hanging fruit - I will gladly implement it.

>  I think generalizations such as moving stuff to HostFacetBase should happen as implementations come in & not before

Basically I agree with you that writing code "for future use" is a bad practice. In this case it's a bit different situation, since we are not _moving_ code, but creating extension points that are already identified as useful. Basically I would say that we should look at all facet related PR's (including puppet) as one, but we have separated them to multiple commits for the sake of granularity of the review.

jlsherrill commented:
> 3) After this, what would be the next step? I think a PR to update Host::Managed to use the kind of stuff that I see on HostFacetBase now would be the wrong course to go vs simply extracting any facet (if there's something simpler than Puppet, that's fine). I think generalizations such as moving stuff to HostFacetBase should happen as implementations come in & not before. Let me know what do you think could go next

@dLobatog there are some extension points that plugins need (such as the rabl and apipie extensions) as there is no other way to extend host json via the api.  Do we need to wait for foreman to migrate to these extension points in order for katello to have access to them?

dLobatog commented:
Merged as 965091345f3ccbd721fdce87c709747deab2f009, thanks @ShimShtein!

domcleal commented:
Please follow up with some documentation, see http://projects.theforeman.org/issues/12630#note-4.  If it's a plugin API (it appears not, see #13417) then on the create plugins page, else a separate wiki page like Pagelets.
